Support ORDER BY and LIMIT on all SELECT-like statements

danhhz opened the pull request:
We were previously accepting these parses, but losing the sort and limit data. This commit adds a generalized sortNode and limitNode for any select-like, but keeps the optimization of passing them down to Select itself.

Closes #4734.
Closes #2691.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4862)

<!-- Reviewable:end -->

paperstreet commented:
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/sort.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fKDviNtKyW5BhyKI:-KC1fKDviNtKyW5BhyKJ:-2075903842) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L40)):*
I'm assuming this isn't bad enough to block beta, so unless there's an objection, I was planning on filing a follow up issue for it

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KC1fDI2huJnKcDUc6D7:-49340100) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
Happily accepting bikesheds on this name

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
cc @petermattis maybe, since radu is out until monday afternoon

tamird commented:
Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/sort.go, line 142 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xi0f40FTOulcUvN3:-KC1xi0f40FTOulcUvN4:1898221714) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L142)):*
i see this changed from a method to a function, but why did it move?

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KC1xBaGWHJw8vv5iyBy:1720289572) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
seems odd for the wrapper to implement this interface - don't you always want a wrapped value? this creates the possibility of a SelectStatement being wrapped or not, which seems like it's going to result in boilerplate code wherever access to the LIMIT or ORDERY BY is needed.

---

*[sql/parser/select.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1wRE7L1tlh_ey4ji9:-KC1wRE7L1tlh_ey4jiA:1692717369) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L85)):*
seems like this wrapper is accepting this responsibility only partially - why does `Select` still have these members at all? the commit message mentions something about an optimization, but I don't totally follow it. Why can't the wrapped value always be passed around and the limit and order by pulled from it?

---

*[sql/testdata/union, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1y7oc4kzAGbjbknq9:-KC1y7oc4kzAGbjbknqA:-1280936503) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/testdata/union#L164)):*
remove

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Nice, though I think we're going to have to bike shed the `sql/parser/select.go` changes and organization.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/sort.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fKDviNtKyW5BhyKI:-KC2BHcRjyTdyY7sbX58:-759598008) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L40)):*
Yes, file an issue. Agreed that this shouldn't block beta.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KC2C34oP75WZEZQpha8:-15037174) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
If we were to follow the naming in `sql.y`, `Select`, `ParenSelect`, `Union` and `Values` should be "select clauses": `SelectClause`, `UnionClause`, `ValueClause`. Which would leave remove for this struct to be named `Select`. There are probably details here that I'm missing and I imagine the renaming would be burdensome. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KC2DQXaydgRB8xR2lta:1700584202)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Okay, no problem.

---

Review status: 10 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/sort.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fKDviNtKyW5BhyKI:-KC2_oUr1fL5Wcs4WImR:-595482790) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L40)):*
https://github.com/cockroachdb/cockroach/issues/4873

---

*[sql/sort.go, line 142 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xi0f40FTOulcUvN3:-KC268qqYF2WIPqUd9l3:1293434800) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L142)):*
It seemed to me like it was in select.go because it was called on a selectNode. When I swapped the selectNode for the numOriginalCols param to make it more general, it seemed no longer specific to select.go. It's now called mostly from this file and once from group.go. Happy to move it back if you like.

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KC2PhyAtzLE7KICJfuV:726883018) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
I was also not entirely convinced of this initially, but it worked out pretty well so I kept it. The big issue is something like `((SELECT 1) LIMIT 1)` which, I think, doesn't parse otherwise. Each () is turned into a `type ParenSelect struct { Select SelectStatement }`. So if SelectOrderLimit isn't a SelectStatement, then we have to do something else there and I'm not sure what.

I think the way that `(*planner) SelectOrderLimit(...)` works eliminates the boilerplate. Willing to revisit if you disagree.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KC2_zw3PSOOS-NrPqZ6:-1344205064) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
parser.Value and value_clause line up. There really isn't a rule directly for Union. parser.Select is more aligned with simple_select than it is with select_clause. I could rename parser.Select to parser.SimpleSelect and use parser.Select for this?

---

*[sql/parser/select.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1wRE7L1tlh_ey4ji9:-KC2VblCZXWih-WU0tmC:-586411157) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L85)):*
I tried removing them from `Select` at one point and decided it was too hard. But I just went back and checked it again and it was fine. *shrug* Good call.

The "optimization" comment was because I was told to keep select aware of its order because it effected index selection, but now that I look closely, I don't see where that actually happens.

---

*[sql/testdata/union, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1y7oc4kzAGbjbknq9:-KC2GEf0tSiqQrZedu9M:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/testdata/union#L164)):*
Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KC2_trZF5-O8l5H5ipb:-265233951)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/sort.go, line 142 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xi0f40FTOulcUvN3:-KC2gXgjhCJc-zI5nHfU:-2131827181) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L142)):*
Nah, it's fine. Moving it and changing it in the same commit makes it more difficult than necessary to spot the changes. Something to consider in the future.

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KC2gpytzMy8Fhbt4UlC:1759418906) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
More than the boilerplate, this just feels awkward - it seems that the underlying "clause" should always be wrapped in this thing, and leaving in place both paths looks unfinished.

---

*[sql/parser/select.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1wRE7L1tlh_ey4ji9:-KC2h_BAXm-LP7I7PNNH:577426989) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L85)):*
Nice. Does the commit message need an update?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KC3UoVd8eXRhIf1aKYa:197581455) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
`simple_select` is either a `SELECT`, `UNION` or `VALUES`. `select_clause` is `simple_select` plus `select_with_parens`. My thought was that `simple_select` should return a `SelectClause` along with `select_with_parens`.

---

*[sql/parser/select.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1wRE7L1tlh_ey4ji9:-KC3Vgvue3wzmjkv3V9k:924279484) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L85)):*
The optimization is that `planner.selectIndex` takes a desired ordering for a scan. The ordering comes from either the `GROUP BY` or `ORDER BY` expression. There are lots of tests that verify we're performing this optimization during index selection, so as long as you haven't broken the tests you're probably good to go.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
I have a version locally that passes all tests, but I have a couple more questions before I update the review.

cc @RaduBerinde in case you hadn't seen this

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KCIKbqHXmecAQBMTKqx:58114136) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
Okay, will do. Can you gut check the following mapping for me?

```%type <SelectStatement> simple_select select_with_parens select_clause values_clause
%type <*SelectOrderLimit> select_stmt select_no_parens```

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KCIJrMMsHTTeSvFEZUv:-2087682562) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
gorename makes the rename pretty easy (though it doesn't seem to handle godoc, I had to do that by hand). A couple followup questions:
- Should I add the refactor to this review as a separate commit before this one? Merge a separate pull request first? Something else?
- If I rename `sql/parser.{Select,Union,Values}` to have a Clause suffix, I should do the same to the corresponding methods on `sql.planner`, right?
- Is there a better name for SelectStatement? It gets confusing with Select and SelectClause. I've come up with a few ideas, but they were all worse than the current one

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KCIKPJWzIf25aqOJe9a:476192000)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented on sql/parser/walk.go:
```diff
@@ -568,7 +559,34 @@ func (stmt *Select) WalkStmt(v Visitor) Statement {
+	ret := stmt
+	sel, changed := WalkStmt(v, stmt.Select)
+	if changed {
+		if ret == stmt {
```
don't need this if

RaduBerinde commented on sql/select_order_limit.go:
```diff
@@ -0,0 +1,57 @@
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+func (p *planner) SelectOrderLimit(n *parser.SelectOrderLimit, autoCommit bool) (planNode, *roachpb.Error) {
```
I would put this in an existing file (maybe select.go) given that it's just one function.

RaduBerinde commented:
The current diff looks pretty good to me. I posted a couple of minor comments outside reviewable.

Just curious, what happens for something like `(SELECT k FROM kv ORDER By v DESC) ORDER BY k`? (postgres errors out with "multiple ORDER BY clauses not allowed")

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KCIVdwGWPE4u1BK2hU2:1219897216) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
- separate commit for renaming existing stuff would be good
- yeah they seem to match up 1-1 currently

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KCIVUtC8e7OvCuuheak:835435862)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KCI_geEZTHz_pOEwYlB:-1694431765) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
Yeah, a separate commit for the renaming would be nice. I have nothing better than `SelectStatement`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KCIuCd0K0qjoiASrAab:449962793) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
I'd have to page this back in; I haven't been in this code in a while. Go ahead with that, and we'll see how it turns out.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

danhhz committed:
Rename sql/parser.{Select, Union, Values} to have a Clause suffix

In preparation for a commit that uses Select to mean a SelectStatement plus OrderBy and/or Limit

danhhz committed:
Support ORDER BY and LIMIT on all SelectStatements

We were previously accepting these parses, but losing the sort and limit data.
This commit adds a generalized sortNode and limitNode for any SelectStatement,
but keeps the optimization of passing them down to SelectClause itself for index
selection.

Closes #4734.
Closes #2691.

paperstreet commented:
RFAL. This next revision has two commits, the first for the refactor, and the second is the limit/order changes. Dunno how Reviewable is going to handle it.

Good call, Radu. Added a check for multiple ORDER BYs.

---

Review status: 4 of 26 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/sort.go, line 142 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xi0f40FTOulcUvN3:-KCMc-bBA2I10-vqXrs2:1792409077) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/sort.go#L142)):*
Gotcha

---

*[sql/parser/select.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1xBaGWHJw8vv5iyBx:-KCMc6M3gr4GIOrQxkVj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L36)):*
Done.

---

*[sql/parser/select.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1fDI2huJnKcDUc6D6:-KCMc7fh4x9dCtboiqjb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L43)):*
Done.

---

*[sql/parser/select.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1wRE7L1tlh_ey4ji9:-KCMccxEghmxzTIFYy_l:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91e03fdbb3ff557847b80efc8e6c627f5ab0cebc/sql/parser/select.go#L85)):*
Done.

---

*[sql/parser/walk.go, line 585 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1f8QMFF4oum8-1p6q-r2-585:-KCMeEdKN9-pEJYSln_K:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3966de2dc656e7ef13fdcc6f778daece40f219e/sql/parser/walk.go#L585)):*
Done.

---

*[sql/select_order_limit.go, line 24 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1f8QMFF4oum8-1p6y-r2-24:-KCMbvisR8EuAM_yhxbs:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3966de2dc656e7ef13fdcc6f778daece40f219e/sql/select_order_limit.go#L24)):*
Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KCMbpZXvVX1ysTfTvUw:738279806)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM but I didn't review that closely. I'd wait for @raduberinde's approval.

---

Reviewed 23 of 23 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/walk.go, line 585 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1f8QMFF4oum8-1p6q-r2-585:-KCMumP5QHUPWRzn2ym8:-92364547) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3966de2dc656e7ef13fdcc6f778daece40f219e/sql/parser/walk.go#L585)):*
doesn't look like this changed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KCMv4MRUhTRVD3Ylwak:1073733077)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/parser/walk.go, line 585 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KC1f8QMFF4oum8-1p6q-r2-585:-KCMw1PTPlr06NgXWu8e:279199428) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3966de2dc656e7ef13fdcc6f778daece40f219e/sql/parser/walk.go#L585)):*
It's an artifact of the diffing. https://github.com/paperstreet/cockroach/commit/18bf3404bd3cd6542b32a587cf81ab0f256ce71a#diff-7dcea9904f292487019a7f81fe41fb56R521 is where it was fixed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for the update, LGTM!

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/select.go, line 203 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KCNNrY2A4Q0wEIApQLw:-KCNNrY3lSLso6nbq9hs:-709325443) ([raw file](https://github.com/cockroachdb/cockroach/blob/18bf3404bd3cd6542b32a587cf81ab0f256ce71a/sql/select.go#L203)):*
maybe mention that this is only used when there is no order or limit clause

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-:-KCNPOsEXM9K9-HwRhYP:-1251216360)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Review status: 24 of 26 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/select.go, line 203 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862#-KCNNrY2A4Q0wEIApQLw:-KCR8WX5mkxMxUunGGLI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/18bf3404bd3cd6542b32a587cf81ab0f256ce71a/sql/select.go#L203)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4862)*
<!-- Sent from Reviewable.io -->
